File: BOGUdanyo/src/main/java/com/backend3rd/BOGUdanyo/service/RouteAccidentService.java
package com.backend3rd.BOGUdanyo.service;

import com.backend3rd.BOGUdanyo.api.KakaoRestApiHelper;
import com.backend3rd.BOGUdanyo.entity.AccidentArea;
import com.backend3rd.BOGUdanyo.repository.RouteAccidentRepository;
import lombok.RequiredArgsConstructor;
import org.json.simple.parser.ParseException;
import org.springframework.context.annotation.PropertySource;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@PropertySource(value={"classpath:application.properties"})
@RequiredArgsConstructor
public class RouteAccidentService {
    final KakaoRestApiHelper kakaoRestApiHelper;

    final RouteAccidentRepository accidentRouteRepository;

    public List<AccidentArea> getRouteAccident(String start, String goal) throws ParseException {
        float [] startC = kakaoRestApiHelper.getKakaoApiFromAddress(start);
        float [] goalC = kakaoRestApiHelper.getKakaoApiFromAddress(goal);

        float middleLon = (startC[0] + goalC[0])/2;
        float middleLat = (startC[1] + goalC[1])/2;

        double radius = distance(middleLat, middleLon, startC[1], startC[0]);

        List<AccidentArea> result = accidentRouteRepository.findByRegion(middleLat, middleLon, radius);

        return result;
    }

    private static double distance(double lat1, double lon1, double lat2, double lon2){
        double theta = lon1 - lon2;
        double dist = (Math.sin(deg2rad(lat1))* Math.sin(deg2rad(lat2)) + Math.cos(deg2rad(lat1))*Math.cos(deg2rad(lat2))*Math.cos(deg2rad(theta)));
        dist = Math.acos(dist);
        dist = rad2deg(dist);
        dist = dist * 60*1.1515*1609.344;

        return dist; //단위 meter
    }

    //10진수를 radian(라디안)으로 변환
    private static double deg2rad(double deg){
        return deg * Math.PI/180.0;
    }
    //radian(라디안)을 10진수로 변환
    private static double rad2deg(double rad){
        return rad * 180 / Math.PI;
    }
}

File: BOGUdanyoKT/src/main/kotlin/com/backend3rd/BOGUdanyoKT/service/RouteAccidentService.kt
package com.backend3rd.BOGUdanyoKT.service

import com.backend3rd.BOGUdanyoKT.api.KakaoRestApiHelper
import com.backend3rd.BOGUdanyoKT.entity.AccidentArea
import com.backend3rd.BOGUdanyoKT.repository.RouteAccidentRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import kotlin.math.PI
import kotlin.math.acos
import kotlin.math.cos
import kotlin.math.sin

@Service
class RouteAccidentService (
        @Autowired val kakaoRestApiHelper: KakaoRestApiHelper,
        @Autowired val routeAccidentRepository: RouteAccidentRepository
        ) {

        fun getRoutAccident(
                start: String, goal: String): List<AccidentArea> {
                var startC = kakaoRestApiHelper.getKakaoApiFromAddress(start)
                var goalC = kakaoRestApiHelper.getKakaoApiFromAddress(goal)

                var middleLat = (startC[1] + goalC[1]) / 2
                var middleLon = (startC[0] + goalC[0]) / 2

                var radius = distance(middleLat, middleLon, startC[1], startC[0])

                return routeAccidentRepository.findByRegion(middleLat, middleLon, radius)
        }

        fun distance(lat1: Float, lon1: Float, lat2: Float, lon2: Float): Double{
                var lat1d = lat1.toDouble()
                var lon1d = lon1.toDouble()
                var lat2d = lat2.toDouble()
                var lon2d = lon2.toDouble()

                var theta = lon1d - lon2d
                var dist = (sin(deg2rad(lat1d)) * sin(deg2rad(lat2d)) + cos(deg2rad(lat1d)) * cos(deg2rad(lat2d)) * cos(deg2rad(theta)))
                dist = acos(dist)
                dist = rad2deg(dist)
                dist *= 60 * 1.1515 * 1609.344

                return dist
        }

        fun deg2rad(deg: Double): Double{ return deg * PI / 180.0}

        fun rad2deg(rad: Double): Double{ return rad * 180 / PI}
}